src/ports/postgres/modules/knn/knn.sql_in

```diff
@@ -0,0 +1,165 @@
+    PythonFunction(knn, knn, knn_validate_src)
+$$ LANGUAGE plpythonu
+m4_ifdef(`__HAS_FUNCTION_PROPERTIES__', `READS SQL DATA', `');
+
+
+CREATE OR REPLACE FUNCTION MADLIB_SCHEMA.knn(
+    arg1 VARCHAR
+) RETURNS VOID AS $$
+BEGIN
+    IF arg1 = 'help' THEN
+	RAISE NOTICE 'You need to enter following arguments in order:
+	Argument 1: Training data table having training features as vector column and labels
+	Argument 2: Name of column having feature vectors in training data table
+	Argument 3: Name of column having actual label/vlaue for corresponding feature vector in training data table
+	Argument 4: Test data table having features as vector column. Id of features is mandatory
+	Argument 5: Name of column having feature vectors in test data table
+	Argument 6: Name of column having feature vector Ids in test data table
+	Argument 7: Name of output table
+	Argument 8: c for classification task, r for regression task
+	Argument 9: value of k. Default will go as 1';
+    END IF;
+END;
+$$ LANGUAGE plpgsql VOLATILE
+m4_ifdef(`__HAS_FUNCTION_PROPERTIES__', `READS SQL DATA', `');
+
+CREATE OR REPLACE FUNCTION MADLIB_SCHEMA.knn(
+) RETURNS VOID AS $$
+BEGIN    
+    EXECUTE $sql$ select * from MADLIB_SCHEMA.knn('help') $sql$;
+END;
+$$ LANGUAGE plpgsql VOLATILE
+m4_ifdef(`__HAS_FUNCTION_PROPERTIES__', `READS SQL DATA', `');
+
+
+CREATE OR REPLACE FUNCTION MADLIB_SCHEMA.knn(
+    point_source VARCHAR,
+    point_column_name VARCHAR,
+    label_column_name VARCHAR,
+    test_source VARCHAR,
+    test_column_name VARCHAR,
+    id_column_name VARCHAR,
+    output_table VARCHAR,
+    operation VARCHAR,
+    k INTEGER
+) RETURNS VARCHAR AS $$
+DECLARE
+    class_test_source REGCLASS;
+    class_point_source REGCLASS;
+    l FLOAT;
+    id INTEGER;
+    vector DOUBLE PRECISION[];
+    cur_pid integer;
+    theResult MADLIB_SCHEMA.knn_result;
+    r MADLIB_SCHEMA.test_table_spec;
+    oldClientMinMessages VARCHAR;
+    returnstring VARCHAR;
+BEGIN
+    oldClientMinMessages :=
+        (SELECT setting FROM pg_settings WHERE name = 'client_min_messages');
+    EXECUTE 'SET client_min_messages TO warning';
+    PERFORM MADLIB_SCHEMA.__knn_validate_src(test_source);
+    PERFORM MADLIB_SCHEMA.__knn_validate_src(point_source);
+    class_test_source := test_source;
+    class_point_source := point_source;
+    --checks
+    IF (k <= 0) THEN
+        RAISE EXCEPTION 'KNN error: Number of neighbors k must be a positive integer.';
+    END IF;
+    IF (operation != 'c' AND operation != 'r') THEN
+        RAISE EXCEPTION 'KNN error: put r for regression OR c for classification.';
+    END IF;
+    PERFORM MADLIB_SCHEMA.create_schema_pg_temp();
+    
+    EXECUTE format('DROP TABLE IF EXISTS %I',output_table);
+    EXECUTE format('CREATE TABLE %I(%I integer, %I DOUBLE PRECISION[], predlabel float)',output_table,id_column_name,test_column_name);
+	
+    
+    FOR r IN EXECUTE format('SELECT %I,%I FROM %I', id_column_name, test_column_name, test_source)
+    LOOP
```

Comment:
This loop forces us to scan the table multiple times which is very costly. We might be able to collapse this into a single level of sql calls. For example, here is a code that finds the 2 closest points (ids and distances) for every test point (assuming you are using the tables from the test code):
`
select * from (
select row_number() over (partition by test_id order by dist) as r, x.*
from (select train.id as tr_id,test.id as test_id,
madlib.squared_dist_norm2(train.data,test.data) as dist
from knn_train_data as train, knn_test_data as test)x)y
where y.r <= 2;`